README.md
Real-Debrid
===========

An unobtrusive Google Chrome extension for Real-Debrid. The extension allows you to download a link or selection of links using the right-click (context) menu.

[Real-Debrid in the Chrome Web Store](https://chrome.google.com/webstore/detail/real-debrid/llhbijccmpenbpkblhmeeneeaangebej)

## Screenshots

![Screenshot](http://i.imgur.com/hMwoToj.png)

![Options](http://i.imgur.com/zgxrlnF.png)

## FAQ

##### Don't I need to enter my login?
No, the extension does not need your credentials. Just make sure you're logged in on real-debrid.com.

##### Do I have to wait until the download starts?
Once you've clicked on *Download with Real-Debrid* you can navigate away from the current page. Your download will start as soon as the link is unrestricted.

## License
[Apache v2 License](https://github.com/JDevlieghere/Real-Debrid/blob/master/LICENSE)


CHANGELOG.md
## 1.10.x

Features:

 - Material design
 - Popup now allows you to paste link and view download history
 - Better notifications
 - Support for unrestricting folders

Bugfixes:

 - Fixed redirection to RD torrents page after adding a magnet

## 1.9.x

Features:

 - Support new Real-Debrid API
 - Authentication using API token (https://real-debrid.com/apitoken)
 - Upgraded to jQuery 3
 - Temporarily removed warnings for limited hosters

## 1.8.x

Features:

 - Torrent support (SublimePotato)
 - Improved save animation (StefanPante)

## 1.7.x

Features:

 - Recent downloads card
 - Account information card
 - Limited hosters card
 - Browser icon added (upper right corner)
 - Bypass native Chrome downloads

Bugfixes:

 - Used percentages fixed
 - Multiple instances of the options page fixed
 - Install bug fixed

## 1.6.x

Features:

 - Options page
 - Configurable warnings

Bugfixes:

 - New installer code using chrome API
 - Typo fixed (Thanks to Nele)
 - Multiple instances of context menu fixed

## 1.5.x

Features:

 - "Extension updated" links to the changelog
 - Download current page's url when no link or text selected
 - Notification when reached 75% of quota for limited hosters
 - Notification when only 7 days left of premium

Bugfixes:

 - Persistent storage of warnings
 - Handle case where account information is requested when not logged in
 - No "extension updated" notification in case of bugfix


Gruntfile.js
module.exports = function(grunt) {

    grunt.initConfig({
        compress: {
            main: {
                options: {
                    archive: 'dist/dist.zip'
                },
                files: [{
                    expand: true,
                    cwd: 'src/',
                    src: ['**/*']
                }]
            }
        },
        jshint: {
            all: ['Gruntfile.js', 'src/js/*.js']
        },
        jsbeautifier: {
            files: ['Gruntfile.js', 'src/js/*.js', 'src/html/*.html'],
            options: {}
        },
        sass: {
            dist: {
                files: [{
                    expand: true,
                    cwd: 'src/scss',
                    src: ['*.scss'],
                    dest: 'src/css',
                    ext: '.css'
                }]
            }
        },
        watch: {
            sass: {
                files: 'src/scss/*.scss',
                tasks: ['sass']
            },
            js: {
                files: 'src/js/*.js',
                tasks: ['jshint', 'jsbeautifier'],
                options: {
                    spawn: false
                }
            }
        }
    });

    grunt.loadNpmTasks('grunt-contrib-compress');
    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-jsbeautifier');
    grunt.loadNpmTasks('grunt-contrib-sass');
    grunt.loadNpmTasks('grunt-contrib-watch');

    grunt.registerTask('default', [
        'sass',
        'jshint',
        'jsbeautifier'
    ]);

    grunt.registerTask('build', [
        'sass',
        'jshint',
        'jsbeautifier',
        'compress'
    ]);

    grunt.registerTask('develop', [
        'sass',
        'watch'
    ]);

};


src/js/popup.js
var pageIdentifier = "popup";

$("#options").click(function() {
    var optionsUrl = chrome.extension.getURL("html/options.html");
    chrome.tabs.query({
        url: optionsUrl
    }, function(tabs) {
        chrome.tabs.create({
            url: optionsUrl
        });
        window.close();
    });
});

$(".downloadPage").click(function() {
    chrome.tabs.getSelected(null, function(tab) {
        chrome.runtime.sendMessage({
            url: tab.url
        }, function() {});
        window.close();
    });
});

$(".unrestrictUrl").click(function() {
    var unrestrictUrl = $("input#unrestrictUrl").val();
    var regex = new RegExp(/((([A-Za-z]{3,9}:(?:\/\/)?)(?:[-;:&=\+\$,\w]+@)?[A-Za-z0-9.-]+|(?:www.|[-;:&=\+\$,\w]+@)[A-Za-z0-9.-]+)((?:\/[\+~%\/.\w-_]*)?\??(?:[-\+=&;%@.\w_]*)#?(?:[\w]*))?)/);
    if (unrestrictUrl.match(regex)) {
        chrome.tabs.getSelected(null, function(tab) {
            chrome.runtime.sendMessage({
                url: unrestrictUrl
            }, function() {});
            window.close();
        });
    } else {
        $('#unrestrictUrl').parent().addClass("is-invalid");
    }
});

var load = function() {
    getDownloads();
};

$(document).ready(load);


src/js/options.js
var URL_TRAFFIC = "https://api.real-debrid.com/rest/1.0/traffic";
var URL_USER = "https://api.real-debrid.com/rest/1.0/user";
var URL_TORRENT_HOSTERS = "https://api.real-debrid.com/rest/1.0/torrents/availableHosts";
var pageIdentifier = "options";

var load = function() {
    chrome.storage.sync.get({
        'apiKey': ''
    }, function(result) {
        $("input#apiKey").val(result.apiKey).parent().addClass("is-dirty");
    });

    chrome.storage.sync.get({
        'warningPercentage': 75
    }, function(result) {
        $("input#warningPercentage").val(result.warningPercentage).parent().addClass("is-dirty");
    });

    chrome.storage.sync.get({
        'warningDays': 7
    }, function(result) {
        $("input#warningDays").val(result.warningDays).parent().addClass("is-dirty");
    });

    chrome.storage.sync.get({
        'bypassNativeDl': false
    }, function(result) {
        if (result.bypassNativeDl) {
            $("#bypassNativeDlLabel").get(0).MaterialSwitch.on();
        }
    });

    chrome.storage.sync.get({
        'splittingSize': 50
    }, function(result) {
        $("input#splittingSize").val(result.splittingSize).parent().addClass("is-dirty");
    });

    chrome.storage.sync.get({
        'torrentHost': "1fichier.com"
    }, function(result) {
        getTorrentHosters(function() {
            $('label[hoster="' + result.torrentHost + '"]')[0].MaterialRadio.check();
        });
    });

    var version = chrome.runtime.getManifest().version;
    $("#version").html(version);

    // A fix for first time setup
    $("section.login-required").children().show();
    $("section.login-required h6").hide();

    getDownloads();
    // getTraffic();
    getAccount();
};

$(document).ready(load);

$(function() {
    var currentPosition = 0;
    var target = 0;
    var header = 0;
    $(".mdl-layout__content").bind("scroll", function() {
        currentPosition = $(".mdl-layout__content").scrollTop();
    });
    $(".mdl-navigation__link").click(function() {
        header = $(".mdl-layout__header").height();
        target = $("#" + this.id + "Card").offset();
        target = target.top - header - 10 + currentPosition;
        scroll(target);
    });
});

function scroll(target) {
    $(".mdl-layout__content").stop().animate({
        scrollTop: target
    }, "slow");
    return false;
}

$("button#save").click(function() {
    var apiKey = $("input#apiKey").val();
    var warningPercentage = $("input#warningPercentage").val();
    var warningDays = $("input#warningDays").val();
    var bypassNativeDl = $("input#bypassNativeDl").parent().hasClass("is-checked");
    var splittingSize = $("input#splittingSize").val();
    var torrentHost = $("input.torrentHost").parent(".is-checked").attr("hoster"); // TODO: Find a better way to read this

    chrome.storage.sync.set({
        'apiKey': apiKey
    });

    if (warningDays && warningPercentage) {
        chrome.storage.sync.set({
            'warningPercentage': warningPercentage
        });

        chrome.storage.sync.set({
            'warningDays': warningDays
        });

        chrome.storage.sync.set({
            'bypassNativeDl': bypassNativeDl
        });

        chrome.storage.sync.set({
            'splittingSize': splittingSize
        });

        chrome.storage.sync.set({
            'torrentHost': torrentHost
        });

        success();
        load();
    } else {
        failure();
    }
});

function getAccount(callback) {
    api(URL_USER, function(data) {
        var expirationDate = moment(data.expiration).format('D MMMM YYYY, H:mm') + " (" + moment(data.expiration, "YYYYMMDD").fromNow() + ")";
        $("#avatar").html('<img src="' + data.avatar + '">');
        $("#username").html(data.username);
        $("#email").html(data.email);
        $("#points").html(data.points);
        $("#expiration").html(expirationDate);
    });
}

// function getTraffic(callback) {
//     api(URL_TRAFFIC, function(data) {
//         if (data.error) {
//             loginRequired();
//         } else {
//             var html = '';
//             var hosters = Object.keys(data);
//             for (var i = 0; i < hosters.length; i++) {
//                 var hoster = data[hosters[i]];
//                 if (hoster.left && hoster.limit) {
//                     var used = Math.round((1 - (parseFloat(hoster.left) / parseFloat(hoster.limit))) * 100);
//                     html += '<div class="row">';
//                     html += '<div class="key">' + hosters[i] + '</div>';
//                     html += '<div class="value">';
//                     html += '<div class="progress">';
//                     html += '<div class="progress-bar" role="progressbar" aria-valuenow="' + used + '" aria-valuemin="0" aria-valuemax="100" style="width: ' + used + '%;">';
//                     html += used + '%';
//                     html += '</div>';
//                     html += '</div>';
//                     html += '</div>';
//                     html += '</div>';
//                 }
//             }
//             $('#hosters').append(html);
//             callback();
//         }
//     });
// }

function getTorrentHosters(callback) {
    api(URL_TORRENT_HOSTERS, function(data) {
        var html = "", // preventing 'undefined' error
            hoster;
        for (var i = 0; i < data.length; i++) {
            hoster = data[i].host.substring(0, data[i].host.indexOf('.'));
            html += `
                    <label class="mdl-radio mdl-js-radio mdl-js-ripple-effect" for="` + hoster + `" hoster="` + data[i].host + `">
                        <input type="radio" class="torrentHost mdl-radio__button" name="hoster" id="` + hoster + `">
                        <span class="mdl-radio__label">` + data[i].host + `</span>
                    </label>
                    <div class="mdl-layout-spacer spacer"></div>
                    `;
        }
        $(".mdl-card #torrentHosters").html(html);
        componentHandler.upgradeAllRegistered(); // Required by MDL
        callback();
    });
}


src/js/functions.js
function bytesToSize(bytes) { // from http://scratch99.com/web-development/javascript/convert-bytes-to-mb-kb/
    var sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB'];
    if (bytes === 0) return 'n/a';
    var i = parseInt(Math.floor(Math.log(bytes) / Math.log(1024)));
    if (i === 0) return bytes + ' ' + sizes[i];
    return (bytes / Math.pow(1024, i)).toFixed(1) + ' ' + sizes[i];
}

function loginRequired() {
    $("section.login-required").children().hide();
    $("section.login-required").append('<h6 class="mdl-color--red-600 mdl-color-text--white"><i class="material-icons">error</i><span>Could not reach Real-Debrid API.</span></h6>');
}

function success() {
    var notification = document.querySelector(".mdl-js-snackbar");
    notification.MaterialSnackbar.showSnackbar({
        message: 'Your options have been saved'
    });
}

function failure() {
    var notification = document.querySelector(".mdl-js-snackbar");
    notification.MaterialSnackbar.showSnackbar({
        message: 'Could not save your options'
    });
}

function getApiKey(callback) {
    chrome.storage.sync.get({
        'apiKey': ''
    }, function(result) {
        callback(result.apiKey);
    });
}

function api(url, callback) {
    getApiKey(function(key) {
        $.ajax({
            type: "GET",
            url: url + "?auth_token=" + key,
            dataType: "json",
            data: {},
            crossDomain: true,
            xhrFields: {
                withCredentials: true
            },
            success: callback,
            error: function() {
                loginRequired();
            }
        });
    });
}

function getDownloads(callback) { // TODO: Find a way to get all hoster icons
    var url = "https://api.real-debrid.com/rest/1.0/downloads";
    api(url, function(data) {
        var html = '',
            createdDate,
            prettySize,
            mime,
            hoster,
            originalUrl,
            unrestrictedUrl;
        if (pageIdentifier == "options") {
            originalUrl = "Original URL";
            unrestrictedUrl = "Unrestricted";
        } else if (pageIdentifier == "popup") {
            originalUrl = "<i class="
            material - icons ">file_download</i>";
            unrestrictedUrl = "<i class="
            material - icons ">cloud_download</i>";
        }
        for (var i = 0; i < data.length; i++) {
            createdDate = moment(data[i].generated).format('D MMM YYYY, H:mm');
            prettySize = bytesToSize(data[i].filesize);
            hoster = data[i].host.substring(0, data[i].host.indexOf('.'));

            if (pageIdentifier == "options") {
                if (!data[i].mimeType) { // Sometimes RD API would return null values
                    data[i].mimeType = "";
                }
                switch (data[i].mimeType) {
                    case (data[i].mimeType.match(/mp4|matroska|wmv|avi|flv/) || {}).input:
                        mime = "video";
                        break;
                    case (data[i].mimeType.match(/audio|mp3|flac|wav|ogg|aac/) || {}).input:
                        mime = "audio";
                        break;
                    case (data[i].mimeType.match(/jpg|png|gif/) || {}).input:
                        mime = "pic";
                        break;
                    case (data[i].mimeType.match(/zip|rar/) || {}).input:
                        mime = "zip";
                        break;
                    case (data[i].mimeType.match(/pdf/) || {}).input:
                        mime = "pdf";
                        break;
                    default:
                        mime = "default";
                        break;
                }
            }
            html += `<div class="recentCards mdl-card mdl-shadow--2dp mdl-cell mdl-cell--12-col">`;
            if (pageIdentifier == "options") {
                html += `
	              <div class="mdl-card__title mdl-color--grey-100 mdl-card--border">
	                <h2 class="mdl-card__title-text"><img src="/icons/mime/` + mime + `.png" title="` + data[i].mimeType + `"></h2>
	              </div>
		  		`;
            }
            html += `
              <div class="mdl-card__supporting-text mdl-color-text--grey-600">
                  <div class="filename"><a>` + data[i].filename + `</a> <small>(` + prettySize + `)</small></div>
                  <small>` + createdDate + `</small>
                  <div class="mdl-layout-spacer spacer"></div>
				  <div class="recentLinks">
	                  <a href="` + data[i].link + `" class="mdl-button mdl-button--colored mdl-js-button mdl-js-ripple-effect" target="_blank">` + originalUrl + `</a>
	                  <a href="` + data[i].download + `" class="mdl-button mdl-button--colored mdl-js-button mdl-js-ripple-effect" target="_blank">` + unrestrictedUrl + `</a>
				  </div>
			  </div>
              <div class="mdl-card__menu">
                <a href="http://www.` + data[i].host + `" target="_blank">
                  <span>
                    <img src="https://cdn.realdebrid.xtnetwork.fr/0693/images/hosters/` + hoster + `.png" title="` + data[i].host + `">
                  </span>
                </a>
              </div>
            </div>
        	`;
        }
        $(".recentDownloads section").append(html);
    });
}


src/js/background.js
/*
Copyright 2014 Jonas Devlieghere

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

var URL_CHANGELOG = "https://github.com/JDevlieghere/Real-Debrid/blob/master/CHANGELOG.md";
var URL_MAGNET = "https://api.real-debrid.com/rest/1.0/torrents/addMagnet";
var URL_TOKEN = "https://real-debrid.com/apitoken";
var URL_TRAFFIC = "https://api.real-debrid.com/rest/1.0/traffic";
var URL_UNRESTRICT = "https://api.real-debrid.com/rest/1.0/unrestrict/link";
var URL_UNRESTRICT_FOLDER = "https://api.real-debrid.com/rest/1.0/unrestrict/folder";
var URL_USER = "https://api.real-debrid.com/rest/1.0/user";

var NORMAL_NOTIFICATION_PREFIX = "id_";
var PROGRESS_NOTIFICATION_PREFIX = "prog_";

var is = new Installer();
var nf = new Notifier();
var op = new Options();

var rd;
var dm;

// Register Options Handler
op.addListener(function() {
    rd = new RealDebrid(op.values.warningPercentage,
        op.values.warningDays,
        op.values.splittingSize,
        op.values.torrentHost);
    dm = new DownloadManager(op.values.bypassNativeDl);
    chrome.downloads.onChanged.addListener(dm.changeHandler);
});

// Load Options
op.load();

// Register Chrome Listeners
chrome.runtime.onInstalled.addListener(is.installHandler);
chrome.runtime.onMessage.addListener(function(request, sender, sendResponse) {
    if (request.url) {
        rd.urlHandler(request.url);
    }
});
chrome.notifications.onClicked.addListener(nf.clickHandler);
chrome.notifications.onButtonClicked.addListener(nf.buttonClickHandler);
chrome.storage.onChanged.addListener(op.changeHandler);

// Create Context Menu
chrome.contextMenus.create({
    "title": "Download with Real-Debrid",
    "contexts": ["link", "selection"],
    "onclick": function(info) {
        if (typeof info.selectionText !== "undefined") {
            rd.selectionHandler(info.selectionText);
        } else if (typeof info.linkUrl !== "undefined") {
            rd.urlHandler(info.linkUrl);
        }
    }
});

function Options() {
    this.values = {};
    this.onLoaded = document.createEvent("Event");

    var that = this;

    this.addListener = function(handler) {
        document.addEventListener("onLoaded", handler, false);
    };

    this.isReady = function() {
        var ready = true;
        for (var key in that.values) {
            ready = ready && that.values[key] !== null;
        }
        return ready;
    };

    this.checkReady = function() {
        if (that.isReady()) {
            document.dispatchEvent(that.onLoaded);
        }
    };

    this.changeHandler = function(changes, namespace) {
        var changed = false;
        for (var key in changes) {
            if (that.values[key]) {
                that.values[key] = changes[key].newValue;
                changed = true;
            }
        }
        if (changed) {
            document.dispatchEvent(that.onLoaded);
        }
    };

    this.init = function() {
        that.onLoaded.initEvent("onLoaded", true, true);
        that.values.warningPercentage = null;
        that.values.warningDays = null;
        that.values.bypassNativeDl = null;
        that.values.splittingSize = null;
        that.values.torrentHost = null;
    };

    this.load = function() {
        that.init();
        chrome.storage.sync.get({
            'warningPercentage': 75
        }, function(result) {
            that.values.warningPercentage = result.warningPercentage;
            that.checkReady();
        });

        chrome.storage.sync.get({
            'warningDays': 7
        }, function(result) {
            that.values.warningDays = result.warningDays;
            that.checkReady();
        });

        chrome.storage.sync.get({
            'bypassNativeDl': false
        }, function(result) {
            that.values.bypassNativeDl = result.bypassNativeDl;
            that.checkReady();
        });

        chrome.storage.sync.get({
            'splittingSize': 50
        }, function(result) {
            that.values.splittingSize = result.splittingSize;
            that.checkReady();
        });

        chrome.storage.sync.get({
            'torrentHost': "1fichier.com"
        }, function(result) {
            that.values.torrentHost = result.torrentHost;
            that.checkReady();
        });

    };
}

/* RealDebrid */
function RealDebrid(warningPercentage, warningDays, splittingSize, torrentHost) {

    this.warnings = [];
    this.warningPercentage = warningPercentage;
    this.warningDays = warningDays;
    this.apiKey = null;

    this.folderSize = 0;

    var that = this;

    chrome.storage.sync.get({
        'apiKey': ''
    }, function(result) {
        that.apiKey = result.apiKey;
    });

    chrome.storage.local.get({
        warnings: []
    }, function(result) {
        that.warnings = result.warnings;
        that.checkAccount();
    });

    this.selectionHandler = function(selection) {
        var urls = selection.split(" ");
        var regex = new RegExp(/((([A-Za-z]{3,9}:(?:\/\/)?)(?:[-;:&=\+\$,\w]+@)?[A-Za-z0-9.-]+|(?:www.|[-;:&=\+\$,\w]+@)[A-Za-z0-9.-]+)((?:\/[\+~%\/.\w-_]*)?\??(?:[-\+=&;%@.\w_]*)#?(?:[\w]*))?)/);
        var invalid;
        $.each(urls, function(index, url) {
            if (url.match(regex)) {
                that.urlHandler(url);
            } else {
                invalid = true;
            }
        });
        if (invalid) {
            nf.error("Selected text doesn't seem to contain valid download URL(s)");
        }
    };

    this.urlHandler = function(url) {
        nf.progress(++nf.progressNotificationId, "Downloading...", "Getting URL", 0);
        var regex = new RegExp(/(\/folder\/)/ig);
        if (url.lastIndexOf("magnet:", 0) === 0) {
            that.handleMagnet(url, function(result) {
                if (result.uri) {
                    chrome.tabs.create({
                        url: "https://real-debrid.com/torrents"
                    });
                } else {
                    nf.error("Error adding magnet");
                }
            });
        } else {
            if (url.match(regex)) {
                that.unrestrictFolder(url, function(result) {
                    nf.progress(nf.progressNotificationId, "Downloading...", url, 25);
                    that.folderSize = result.length;
                    $.each(result, function(index, results) {
                        results = results.substring(0, results.indexOf('"')); // RD API seems to return some extra incorrectly formatted information
                        that.urlHandlerExtension(results);
                    });
                });
            } else {
                that.urlHandlerExtension(url);
            }
        }
    };

    this.urlHandlerExtension = function(url) {
        nf.progress(nf.progressNotificationId, "Downloading...", url, 50);
        that.unrestrict(url, function(result) {
            if (result.download) {
                that.download(result.download);
            } else {
                nf.error("Error adding download");
            }
        });
    };

    this.download = function(data) {
        dm.download(data);
    };

    this.post = function(url, data, callback) {
        $.ajax({
            type: "POST",
            url: url + "?auth_token=" + that.apiKey,
            dataType: "json",
            data: data,
            crossDomain: true,
            xhrFields: {
                withCredentials: true
            },
            success: callback,
            error: function(e) {
                that.handleError(e.responseText);
            }
        });
    };

    this.get = function(url, callback) {
        $.ajax({
            type: "GET",
            url: url + "?auth_token=" + that.apiKey,
            dataType: "json",
            data: {},
            crossDomain: true,
            xhrFields: {
                withCredentials: true
            },
            success: callback,
            error: function(e) {
                that.handleError(e.responseText);
            }
        });
    };

    this.handleError = function(result) {
        var prased_result = $.parseJSON(result);
        switch (prased_result.error_code) {
            case -1:
                nf.error("Real-Debrid internal error");
                break;
            case 8:
            case 9:
                nf.error("Unable to authenticate with Real-Debrid. Please click here to set-up your API key.", nf.openOptions);
                break;
            case 14:
                nf.error("Account locked");
                break;
            case 15:
                nf.error("Account not activated");
                break;
            case 16:
                nf.error("Unsupported hoster");
                break;
            case 17:
                nf.error("Hoster in maintenance");
                break;
            case 18:
                nf.error("Hoster limit reached");
                break;
            case 19:
                nf.error("Hoster temporarily unavailable");
                break;
            case 20:
                nf.error("Hoster not available for free users");
                break;
            case 21:
                nf.error("Too many active downloads");
                break;
            default:
                nf.error("Uh Oh. Real-Debrid API responded with an unknown error. Code #: " + prased_result.error_code);
        }
    };

    this.unrestrict = function(url, callback) {
        that.post(URL_UNRESTRICT, {
            link: url
        }, callback);
    };

    this.unrestrictFolder = function(url, callback) {
        that.post(URL_UNRESTRICT_FOLDER, {
            link: url
        }, callback);
    };

    this.account = function(callback) {
        that.get(URL_USER, callback);
    };

    this.traffic = function(callback) {
        that.get(URL_TRAFFIC, callback);
    };

    this.handleMagnet = function(magnetLink, callback) {
        that.post(URL_MAGNET, {
            magnet: magnetLink,
            split: splittingSize,
            host: torrentHost
        }, callback);
    };

    this.checkPremium = function(data) {
        var today = new Date();
        var expiration = new Date(data.expiration);
        var daysLeft = Math.round(Math.abs((expiration.getTime() - today.getTime()) / (24 * 60 * 60 * 1000)));

        var key = "premium-left";
        var index = that.warnings.indexOf(key);

        if (daysLeft <= that.warningDays && index === -1) {
            nf.info("You have only " + daysLeft + " days left of premium. Click here to change warnings preferences.", nf.openOptions);
            that.storeWarning(key);
        } else if (daysLeft > that.warningDays && index !== -1) {
            that.removeWarning(key);
        }
    };

    this.checkAccount = function() {
        that.account(function(result) {
            that.checkPremium(result);
        });
    };

    this.storeWarning = function(warning) {
        that.warnings.push(warning);
        chrome.storage.local.set({
            warnings: that.warnings
        }, function() {});
    };

    this.removeWarning = function(warning) {
        var index = that.warnings.indexOf(warning);
        that.warnings.splice(index, 1);
        chrome.storage.local.set({
            warnings: that.warnings
        }, function() {});
    };

}

/* Notifier */
function Notifier() {

    this.notificationId = 0;
    this.progressNotificationId = 0;
    this.callbacks = {};
    this.urls = [];

    var that = this;

    this.basic = function(title, text, onClicked) {
        var id = ++that.notificationId;
        var options = {
            iconUrl: "/icons/icon-128.png",
            type: "basic",
            title: title,
            message: text,
            priority: 1
        };
        chrome.notifications.create(NORMAL_NOTIFICATION_PREFIX + id, options, function(notificationId) {
            if (onClicked) {
                that.callbacks[notificationId] = onClicked;
            }
        });
    };

    this.progress = function(id, title, text, progress) {
        var id = id;
        var options = {
            iconUrl: "/icons/icon-128.png",
            type: "progress",
            title: title,
            message: text,
            priority: 1,
            progress: progress
        };

        if (progress == 100) {
            options.buttons = [{
                title: "Copy URL To Clipboard",
                iconUrl: "/icons/copy_18dp_2x.png"
            }];
        }
        if (progress > 0) {
            chrome.notifications.update(PROGRESS_NOTIFICATION_PREFIX + id, options, function() {
                if (progress == 100 && rd.folderSize <= 1) {
                    that.urls.push({
                        id: PROGRESS_NOTIFICATION_PREFIX + id,
                        url: text
                    });
                }
            });
        } else {
            chrome.notifications.create(PROGRESS_NOTIFICATION_PREFIX + id, options);
            nf.urls.push({
                id: PROGRESS_NOTIFICATION_PREFIX + nf.progressNotificationId,
                url: ""
            });
        }
    };

    this.error = function(text, callback) {
        that.basic("Error", text, callback);
        chrome.notifications.clear(PROGRESS_NOTIFICATION_PREFIX + nf.progressNotificationId);
    };

    this.info = function(text, callback) {
        that.basic("Real-Debrid", text, callback);
    };

    this.clickHandler = function(notificationId) {
        if (that.callbacks[notificationId]) {
            that.callbacks[notificationId]();
            delete that.callbacks[notificationId];
        }
    };

    this.buttonClickHandler = function(notificationId, buttonIndex) {
        var clicked = that.urls.find(function(obj) {
            return obj.id == notificationId;
        });
        clipboard.copy(clicked.url);
    };

    this.openOptions = function() {
        var optionsUrl = chrome.extension.getURL("html/options.html");
        chrome.tabs.query({
            url: optionsUrl
        }, function(tabs) {
            if (tabs.length) {
                chrome.tabs.update(tabs[0].id, {
                    active: true
                });
            } else {
                chrome.tabs.create({
                    url: optionsUrl
                });
            }
            window.close();
        });
    };
}

/* Installer */
function Installer() {

    var that = this;

    this.onInstall = function(currVersion) {
        chrome.tabs.create({
            url: "html/options.html"
        }, function() {
            nf.info("Extension installed");
        });
    };

    this.onUpdate = function(prevVersion, currVersion) {
        var prevVersionDigits = prevVersion.split('.');
        var currVersionDigits = currVersion.split('.');
        if (prevVersionDigits.length >= 2 &&
            currVersionDigits.length >= 2 &&
            prevVersionDigits[0] == currVersionDigits[0] &&
            prevVersionDigits[1] == currVersionDigits[1]) {
            console.log("Extension updated (bugfix) to version " + currVersion);
        } else {
            var message = "Extension updated to version " + currVersion + ". Click here to see all changes.";
            nf.info(message, function() {
                chrome.tabs.create({
                    url: URL_CHANGELOG
                }, function() {});
            });
        }
    };

    this.installHandler = function(details) {
        var currVersion = chrome.runtime.getManifest().version;
        if (details.reason == "install") {
            that.onInstall(currVersion);
        } else if (details.reason == "update") {
            that.onUpdate(details.previousVersion, currVersion);
        }
    };
}

/* Download Manager */
function DownloadManager(bypassNativeDl) {

    this.active = [];
    var that = this;

    this.download = function(url) {
        if (!bypassNativeDl) {
            chrome.downloads.download({
                url: url
            }, function(downloadId) {
                that.addToActive(downloadId);
            });
        } else {
            chrome.tabs.create({
                url: url,
                active: false
            });
        }

        if (!rd.folderSize) {
            nf.progress(nf.progressNotificationId, "Success!", url, 100);
        } else if (rd.folderSize == 1) {
            var l = nf.urls.length - 1;
            nf.urls[l].url += url;
            nf.progress(nf.progressNotificationId, "Success!", "All folder files successfully added to download queue", 100);
        } else {
            rd.folderSize--;
            var l = nf.urls.length - 1;
            nf.urls[l].url += url + "\n";
        }
    };

    this.checkComplete = function() {
        if (that.active.length === 0) {
            nf.info("All downloads complete", nf.openOptions);
        }
    };

    this.addToActive = function(id) {
        that.active.push(id);
    };

    this.removeFromActive = function(id) {
        var index = that.active.indexOf(id);
        that.active.splice(index, 1);
    };

    this.changeHandler = function(downloadItemDelta) {
        if (that.active.indexOf(downloadItemDelta.id) > -1 && downloadItemDelta.state) {
            if (downloadItemDelta.state.current == "complete") {
                that.removeFromActive(downloadItemDelta.id);
                that.checkComplete();
            } else if (downloadItemDelta.state.current == "interrupted") {
                that.removeFromActive(downloadItemDelta.id);
            }
            rd.checkAccount();
        }
    };
}
